BinaryEncoding.md

```diff
@@ -1,122 +1,400 @@
+are listed below.
+
+### Memory section
+The memory section declares the size and characteristics of the memory associated with the module.
+A module may contain at most one memory section.
+
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| id = `0x00` | `uint8` | section identifier for memory |
+| min_mem_size | `uint8` | minimize memory size as a power of `2` |
+| max_mem_size | `uint8` | maximum memory size as a power of `2` |
+| exported | `uint8` | `1` if the memory is visible outside the module |
+
+### Signatures section
+The signatures section declares all function signatures that will be used in the module.
+A module may contain at most one signatures section.
+
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| id = `0x01` | `uint8` | section identifier for signatures |
+| count | `varuint32` | count of signature entries to follow | 
+| entries | `signature_entry*` | repeated signature entries as described below |
+
+#### Signature entry
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| param_count | `uint8` | the number of parameters to the function |
+| return_type | `local_type?` | the return type of the function, with `0` indicating no return type |
+| param_types | `local_type*` | the parameter types of the function, with<br>`1` indicating type `i32`<br>`2` indicating type `i64`<br>`3` indicating type `f32`<br>`4` indicating type `f64` |
+
+### Functions section
+The Functions section declares the functions in the module and must be preceded by a [Signatures](#signatures-section) section. A module may contain at most one functions section.
+
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| id = `0x02` | `uint8` | section identifier for functions |
+| count | `varuint32` | count of function entries to follow | 
+| entries | `function_entry*` | repeated function entries as described below |
+
+#### Function Entry
+
+Each function entry describes a function that can be optionally named, imported and/or exported. Non-imported functions
+must contain a function body.
+
+| Field | Type |  Present?  | Description |
+| ----- |  ----- |  ----- |  ----- | 
+| flags | `uint8` | always | flags indicating attributes of a function <br>bit `0` : a name is present<br>bit `1` : the function is an import<br>bit `2` : the function has local variables<br>bit `3` : the function is an export |
+| signature | `uint16` | always | index into the Signature section |
+| name | `uint32` | `flags[0] == 1` | name of the function as an offset within the module |
+| i32 count | `uint16` | `flags[2] == 1` | number of i32 local variables |
+| i64 count | `uint16` | `flags[2] == 1` | number of i64 local variables |
+| f32 count | `uint16` | `flags[2] == 1` | number of f32 local variables |
+| f64 count | `uint16` | `flags[2] == 1` | number of f64 local variables |
+| body size | `uint16` | `flags[0] == 0` | size of function body to follow, in bytes |
+| body | `bytes` | `flags[0] == 0` | function body |
+
```

Comment:
Is there a constraint that a name is required for an import or export? Is a name allowed is neither?

The body_size and body depend on bit 1 (not an import)?